classes/ProductFilter/FilterApplication/FilterValidator.php

```diff
@@ -0,0 +1,272 @@
+        }
+        if (!isset($filter['value'])) {
+            throw new InvalidArgumentException('Filter #' . $index . ' requires a "value" field.');
+        }
+    }
+
+    /**
+     * @throws InvalidArgumentException
+     *
+     * @param $filter
+     * @param int $index
+     *
+     * @return void
+     */
+    protected function mustHaveSeveralValues($filter, int $index): void
+    {
+        if (isset($filter['value'])) {
+            throw new InvalidArgumentException('Filter #' . $index . ' is malformed, a field "values" is excepted instead of "value".');
+        }
+        if (!isset($filter['values'])) {
+            throw new InvalidArgumentException('Filter #' . $index . ' requires a "values" field.');
+        }
+        if (!is_array($filter['values']) || $this->isAssociativeArray($filter['values'])) {
+            throw new InvalidArgumentException('Filter #' . $index . ' is malformed, "values" field need to be an array of value.');
+        }
+    }
+
+    /**
+     * @throws InvalidArgumentException
+     *
+     * @param $filter
+     * @param int $index
+     *
+     * @return void
+     */
+    protected function mustBeBoolean($filter, int $index): void
+    {
+        if (gettype($filter['value']) !== 'boolean') {
+            throw new InvalidArgumentException("Value of filter #$index must be a boolean, " . gettype($filter['value']) . ' provided');
+        }
+    }
+
+    /**
+     * @throws InvalidArgumentException
+     *
+     * @param $filter
+     * @param int $index
+     *
+     * @return void
+     */
+    protected function mustBeString($filter, int $index): void
+    {
+        $checkValueIsString = function ($value) use ($index): void {
+            if (!is_string($value)) {
+                throw new InvalidArgumentException('Value ' . $value . " of filter #$index must be a string.");
+            }
+        };
+
+        // Single Value
+        if (isset($filter['value'])) {
+            $checkValueIsString($filter['value']);
+        }
+
+        // Multiple values
+        if (isset($filter['values'])) {
+            foreach ($filter['values'] as $value) {
+                $checkValueIsString($value);
+            }
+        }
+    }
+
+    /**
+     * @throws InvalidArgumentException
+     *
+     * @param $filter
+     * @param int $index
+     *
+     * @return void
+     */
+    protected function mustBeNumber($filter, int $index): void
+    {
+        $checkValueIsNumber = function ($value) use ($index): void {
+            if (!is_numeric($value)) {
+                throw new InvalidArgumentException('Value ' . $value . " of filter #$index must be a number.");
+            }
+        };
+
+        // Single Value
+        if (isset($filter['value'])) {
+            $checkValueIsNumber($filter['value']);
+        }
+
+        // Multiple values
+        if (isset($filter['values'])) {
```

Comment:
I reckon it's either value or values, right ?
So there's no need to do the second isset